src/query/storages/fuse/src/io/read/block_reader.rs

```diff
@@ -71,6 +119,57 @@ impl OwnerMemory {
     }
 }
 
+#[derive(Clone)]
+pub struct ChunksReader<'a> {
+    bytes: VecDeque<&'a [u8]>,
+}
+
+impl<'a> ChunksReader<'a> {
+    pub fn create(bytes: VecDeque<&'a [u8]>) -> ChunksReader<'a> {
+        ChunksReader::<'a> { bytes }
+    }
+
+    #[allow(clippy::wrong_self_convention)]
+    pub fn to_vec(mut self) -> Vec<u8> {
+        let len = self.bytes.iter().map(|s| s.len()).sum();
+        let mut array = Vec::with_capacity(len);
+
+        while let Some(bytes) = self.bytes.pop_front() {
+            array.extend_from_slice(bytes);
+        }
+
+        array
+    }
+}
+
+impl<'a> Read for ChunksReader<'a> {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        while let Some(first) = self.bytes.front_mut() {
+            match first.read(buf)? {
+                0 if !buf.is_empty() => {
+                    self.bytes.pop_front();
+                }
+                n => return Ok(n),
```

Comment:
buf could be much larger than single chunk of bytes. We need to check `buf.is_empty` and kept writing into it.